File: Freestylecoding.Math.FSharp.Tests/Natural/BitwiseOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public BitwiseAnd() =
    [<Theory>]
    [<InlineData(  0u,  0u, 0u)>]
    [<InlineData(  1u,  0u, 0u)>]
    [<InlineData(  0u,  1u, 0u)>]
    [<InlineData(  1u,  1u, 1u)>]
    [<InlineData( 12u, 10u, 8u)>]
    member public this.Sanity left right expected =
        Assert.Equal( Natural( [expected] ), Natural( [left] ) &&& Natural( [right] ) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.Equal( Natural.Unit, Natural([0xFu; 0x00000101u]) &&& Natural([ 0x00010001u]) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.Equal( Natural.Unit, Natural([0x00010001u]) &&& Natural([0xFu; 0x00000101u]) )

    [<FactAttribute>]
    member public this.Large () =
        Assert.Equal( Natural( [1u; 0u; 0u; 0u] ), Natural( [1u; 1u; 0u; 0u] ) &&& Natural( [1u; 0u; 1u; 0u] ) )

type public BitwiseOr() =
    [<Theory>]
    [<InlineData(  0u,  0u,  0u)>]
    [<InlineData(  1u,  0u,  1u)>]
    [<InlineData(  0u,  1u,  1u)>]
    [<InlineData(  1u,  1u,  1u)>]
    [<InlineData( 12u, 10u, 14u)>]
    member public this.Sanity left right expected =
        Assert.Equal( Natural([expected]), Natural([left]) ||| Natural([right]) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.Equal( Natural([0xFu; 0x10101u]), Natural([0xFu; 0x00000101u]) ||| Natural([ 0x00010001u]) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.Equal( Natural([0xFu; 0x10101u]), Natural([0x00010001u]) ||| Natural([0xFu; 0x00000101u]) )

    [<FactAttribute>]
    member public this.Large () =
        Assert.Equal( Natural( [1u; 1u; 1u; 0u] ), Natural( [1u; 1u; 0u; 0u] ) ||| Natural( [1u; 0u; 1u; 0u] ) )

type public BitwiseXor() =
    [<Theory>]
    [<InlineData(  0u,  0u, 0u)>]
    [<InlineData(  1u,  0u, 1u)>]
    [<InlineData(  0u,  1u, 1u)>]
    [<InlineData(  1u,  1u, 0u)>]
    [<InlineData( 12u, 10u, 6u)>]
    member public this.Sanity left right expected =
        Assert.Equal( Natural([expected]), Natural([left]) ^^^ Natural([right]) )

    [<Fact>]
    member public this.BiggerLeft () =
        Assert.Equal( Natural([0xFu; 0x10100u]), Natural([0xFu; 0x00000101u]) ^^^ Natural([0x00010001u]) )

    [<Fact>]
    member public this.BiggerRight () =
        Assert.Equal( Natural([0xFu; 0x10100u]), Natural([0x00010001u]) ^^^ Natural([0xFu; 0x00000101u]) )

    [<FactAttribute>]
    member public this.Large () =
        Assert.Equal( Natural( [0u; 1u; 1u; 0u] ), Natural( [1u; 1u; 0u; 0u] ) ^^^ Natural( [1u; 0u; 1u; 0u] ) )

type public BitwiseNot() =
    [<Theory>]
    [<InlineData( 0xFFFF_FFFEu, 1u)>]
    [<InlineData( 1u, 0xFFFF_FFFEu)>]
    member public this.Sanity value expected =
        Assert.Equal( Natural([expected]), ~~~ Natural([value]) )

    [<Fact>]
    member public this.Bigger () =
        Assert.Equal( Natural([0xF012_3456u; 0x789A_BCDEu]), ~~~ Natural([0x0FED_CBA9u; 0x8765_4321u]) )

type public LeftShift() =
    [<Theory>]
    [<InlineData(   1u, 1,    2u )>]        // Sanity
    [<InlineData( 0xFu, 2, 0x3Cu )>]   // multiple bits
    member public this.Sanity left right expected =
        Assert.Equal( Natural([expected]), Natural([left]) <<< right )
    
    [<Fact>]
    member public this.Overflow () =
        Assert.Equal( Natural([1u; 0xFFFFFFFEu]), Natural([0xFFFFFFFFu]) <<< 1 )
    
    [<Fact>]
    member public this.MultipleOverflow () =
        Assert.Equal( Natural([0x5u; 0xFFFFFFF8u]), Natural([0xBFFFFFFFu]) <<< 3 )
    
    [<Fact>]
    member public this.OverOneUInt () =
        Assert.Equal( Natural([8u; 0u; 0u]), Natural.Unit <<< 67 )

type public RightShift() =
    [<Theory>]
    [<InlineData(    1u, 1,   0u )>]        // Sanity
    [<InlineData(  0xFu, 2, 0x3u )>]    // multiple bits
    [<InlineData( 0x3Cu, 2, 0xFu )>]   // multiple bits
    member public this.Sanity left right expected =
        Assert.Equal( Natural([expected]), Natural([left]) >>> right )

    [<Fact>]
    member public this.Underflow () =
        Assert.Equal( Natural([0x7FFFFFFFu]), Natural([0xFFFFFFFFu]) >>> 1 )
    
    [<Fact>]
    member public this.MultipleUnderflow () =
        Assert.Equal( Natural([0x1u; 0x5FFFFFFFu]), Natural([0xAu; 0xFFFFFFFFu]) >>> 3 )
    
    [<Fact>]
    member public this.OverOneUInt () =
        Assert.Equal( Natural.Unit, Natural([0x10u; 0u; 0u]) >>> 68 )
    
    [<Fact>]
    member public this.ReduceToZero () =
        Assert.Equal( Natural.Zero, Natural([0x10u; 0u; 0u]) >>> 99 )

File: Freestylecoding.Math.CSharp.Tests/Natural/BitwiseOperators.cs
using Xunit;

namespace Freestylecoding.Math.CSharp.Tests.Natural;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class BitwiseAnd {
	[Theory]
	[InlineData( 0u, 0u, 0u )]
	[InlineData( 1u, 0u, 0u )]
	[InlineData( 0u, 1u, 0u )]
	[InlineData( 1u, 1u, 1u )]
	[InlineData( 12u, 10u, 8u )]
	public void Sanity( uint left, uint right, uint expected ) =>
		Assert.Equal(
			new Natural( expected ),
			new Natural( left ) & new Natural( right )
		);

	[Fact]
	public void BiggerLeft() =>
		Assert.Equal(
			Natural.Unit,
			FListNatural( 0xFu, 0x00000101u ) & new Natural( 0x00010001u )
		);

	[Fact]
	public void BiggerRight() =>
		Assert.Equal(
			Natural.Unit,
			new Natural( 0x00010001u ) & FListNatural( 0xFu, 0x00000101u )
		);

	[Fact]
	public void Large() =>
		Assert.Equal(
			FListNatural( 1u, 0u, 0u, 0u ),
			FListNatural( 1u, 1u, 0u, 0u ) & FListNatural( 1u, 0u, 1u, 0u )
		);
}

public class BitwiseOr {
	[Theory]
	[InlineData( 0u, 0u, 0u )]
	[InlineData( 1u, 0u, 1u )]
	[InlineData( 0u, 1u, 1u )]
	[InlineData( 1u, 1u, 1u )]
	[InlineData( 12u, 10u, 14u )]
	public void Sanity( uint left, uint right, uint expected ) =>
		Assert.Equal(
			new Natural( expected ),
			new Natural( left ) | new Natural( right )
		);

	[Fact]

	public void BiggerLeft() =>
		Assert.Equal(
			FListNatural( 0xFu, 0x10101u ),
			FListNatural( 0xFu, 0x00000101u ) | new Natural( 0x00010001u )
		);

	[Fact]
	public void BiggerRight() =>
		Assert.Equal(
			FListNatural( 0xFu, 0x10101u ),
			new Natural( 0x00010001u ) | FListNatural( 0xFu, 0x00000101u )
		);

	[Fact]
	public void Large() =>
		Assert.Equal(
			FListNatural( 1u, 1u, 1u, 0u ),
			FListNatural( 1u, 1u, 0u, 0u ) | FListNatural( 1u, 0u, 1u, 0u )
		);
}

public class BitwiseXor {
	[Theory]
	[InlineData( 0u, 0u, 0u )]
	[InlineData( 1u, 0u, 1u )]
	[InlineData( 0u, 1u, 1u )]
	[InlineData( 1u, 1u, 0u )]
	[InlineData( 12u, 10u, 6u )]
	public void Sanity( uint left, uint right, uint expected ) =>
		Assert.Equal(
			new Natural( expected ),
			new Natural( left ) ^ new Natural( right )
		);

	[Fact]
	public void BiggerLeft() =>
		Assert.Equal(
			FListNatural( 0xFu, 0x10100u ),
			FListNatural( 0xFu, 0x00000101u ) ^ new Natural( 0x00010001u )
		);

	[Fact]
	public void BiggerRight() =>
		Assert.Equal(
			FListNatural( 0xFu, 0x10100u ),
			new Natural( 0x00010001u ) ^ FListNatural( 0xFu, 0x00000101u )
		);

	[Fact]
	public void Large() =>
		Assert.Equal(
			FListNatural( 0u, 1u, 1u, 0u ),
			FListNatural( 1u, 1u, 0u, 0u ) ^ FListNatural( 1u, 0u, 1u, 0u )
		);
}

public class BitwiseNot {
	[Theory]
	[InlineData( 0xFFFF_FFFEu, 1u )]
	[InlineData( 1u, 0xFFFF_FFFEu )]
	public void Sanity( uint value, uint expected ) =>
		Assert.Equal( new Natural( expected ), ~new Natural( value ) );

	[Fact]
	public void Bigger() =>
		Assert.Equal(
			FListNatural( 0xF012_3456u, 0x789A_BCDEu ),
			~FListNatural( 0x0FED_CBA9u, 0x8765_4321u )
		);
}

public class LeftShift {
	[Theory]
	[InlineData( 1u, 1, 2u )]   // Sanity
	[InlineData( 0xFu, 2, 0x3Cu )]  // multiple bits
	public void Sanity( uint left, int right, uint expected ) =>
		Assert.Equal( new Natural( expected ), new Natural( left ) << right );

	[Fact]
	public void Overflow() =>
		Assert.Equal( FListNatural( 1u, 0xFFFFFFFEu ), new Natural( 0xFFFFFFFFu ) << 1 );

	[Fact]
	public void MultipleOverflow() =>
		Assert.Equal( FListNatural( 0x5u, 0xFFFFFFF8u ), new Natural( 0xBFFFFFFFu ) << 3 );

	[Fact]
	public void OverOneUInt() =>
		Assert.Equal( FListNatural( 8u, 0u, 0u ), Natural.Unit << 67 );
}

public class RightShift {
	[Theory]
	[InlineData( 1u, 1, 0u )]   // Sanity
	[InlineData( 0xFu, 2, 0x3u )]   // multiple bits
	[InlineData( 0x3Cu, 2, 0xFu )]  // multiple bits
	public void Sanity( uint left, int right, uint expected ) =>
		Assert.Equal( new Natural( expected ), new Natural( left ) >> right );

	[Fact]
	public void Underflow() =>
		Assert.Equal( new Natural( 0x7FFFFFFFu ), new Natural( 0xFFFFFFFFu ) >> 1 );

	[Fact]
	public void MultipleUnderflow() =>
		Assert.Equal( FListNatural( 0x1u, 0x5FFFFFFFu ), FListNatural( 0xAu, 0xFFFFFFFFu ) >> 3 );

	[Fact]
	public void OverOneUInt() =>
		Assert.Equal( Natural.Unit, FListNatural( 0x10u, 0u, 0u ) >> 68 );

	[Fact]
	public void ReduceToZero() =>
		Assert.Equal( Natural.Zero, FListNatural( 0x10u, 0u, 0u ) >> 99 );
}